packages/nx/src/command-line/graph/inputs-utils.ts

```diff
@@ -0,0 +1,292 @@
+ */
+import { minimatch } from 'minimatch';
+
+import { ProjectGraphProjectNode } from '../../config/project-graph';
+import { FileData } from '../../config/project-graph';
+import { InputDefinition } from '../../config/workspace-json-project-json';
+import { getRootTsConfigPath } from '../../plugins/js/utils/typescript';
+import {
+  expandNamedInput,
+  extractPatternsFromFileSets,
+  filterUsingGlobPatterns,
+} from '../../hasher/task-hasher';
+import type { ProjectGraphClientResponse } from './graph';
+
+export interface ExpandedInputs {
+  general?: { [plan: string]: string[] };
+  [projectName: string]: { [plan: string]: string[] };
+  external?: { [plan: string]: string[] };
+}
+
+/**
+ * This function is used to expand the inputs for the graph command.
+ * It categorizes the inputs and expand them into file paths.
+ * @param plans e.g. ['{workspaceRoot}/**', 'app:src/**']
+ * @param allWorkspaceFiles
+ * @param depGraphClientResponse
+ * @returns a record of expanded inputs. e.g. {general: [{file: 'apps/app/src/main.ts', plan: '{workspaceRoot}/apps/app/src/main.ts'}], app: [{file: 'src/main.ts', plan: 'app:src/main.ts'}], external: [{file: 'node_modules/some-lib/index.ts', plan: 'node_modules/some-lib/index.ts'}]
+ */
+export function expandInputs(
+  plans: string[],
+  allWorkspaceFiles: FileData[],
+  depGraphClientResponse: ProjectGraphClientResponse,
+  taskId: string
+): ExpandedInputs {
+  const projectNames = depGraphClientResponse.projects.map((p) => p.name);
+
+  const {
+    workspaceRootInputs,
+    projectRootInputs,
+    externalInputs,
+    otherInputs,
+  } = categorizeInputs(plans, projectNames, taskId);
+
+  const general = {
+    ...getExpandWorkspaceRootInputs(workspaceRootInputs, allWorkspaceFiles),
+    ...getExpandOtherInputs(otherInputs),
+  };
+  const project = getExpandProjectInputs(
+    projectRootInputs,
+    depGraphClientResponse
+  );
+  const external = externalInputs.reduce((acc, plan) => {
+    acc[plan] = [plan];
+    return acc;
+  }, {});
+
+  return {
+    general: Object.values(general).length ? general : undefined,
+    external: Object.values(external).length ? external : undefined,
+    ...project,
+  };
+}
+
+/**
+ * This function is used to categorize the inputs into different categories
+ */
+function categorizeInputs(
+  inputs: string[],
+  projectNames: string[],
+  taskId: string
+): {
+  workspaceRootInputs: string[];
+  projectRootInputs: string[];
+  externalInputs: string[];
+  otherInputs: string[];
+} {
+  const workspaceRootInputs: string[] = [];
+  const projectRootInputs: string[] = [];
+  const externalInputs: string[] = [];
+  const otherInputs: string[] = [];
+  inputs.forEach((input) => {
+    if (input.startsWith('{workspaceRoot}')) {
+      workspaceRootInputs.push(input);
+      return;
+    }
+    if (input.startsWith('{projectRoot}')) {
+      const projectName = taskId.split(':')[0];
+      projectRootInputs.push(`${projectName}:${input}`);
```

Comment:
this is a nit but is it really the best way to add the `projectName` to the string only to replace it again later? Shouldn't we instead use an object or some other data structure vs. a string?
Just a thought, feel free to ignore :P

Reply:
changed it be an object